### PR TITLE
Added new errors. (Issue #68)

### LINKED ISSUE #68: More informative errors
Currently, we are using the `anyhow` crate for error handling because it is very easy to use - it exposes one error type `anyhow::Error`, which is a catch-all for all types implementing `std::error::Error`.

This approach has a big drawback - while `anyhow::Error` can be associated with a descriptive error message so that it is easy to understand by a human, it's not really possible to differentiate different kinds of failure in the code - e.g. distinguish between query timeout and connection close. This kind of information will not only be useful for client programs, but for driver internals, too - for example, see https://github.com/psarna/scylla-rust-driver/pull/59#discussion_r511469036

We should gradually move away from the `anyhow::Error` type and write meaningful error types ourselves. I'd like to suggest the [thiserror](https://docs.rs/thiserror/1.0.21/thiserror/) crate - it greatly simplifies writing custom error types.

### PR DESCRIPTION
**Pull request by asledz**
Added new typed errors using thiserror.
Deleted anyhow errors.
Fixes #68

### CODE REVIEW

**Comment by asledz** [comment]
@psarna @piodul please review

**Comment by piodul** [comment]
While this is definitely a step in the right direction - now it is easy to match on different kinds of errors in the code - I think the error types could be organized better.

First, it looks like now we have a single, catch-all type which is returned from all methods of `Session` - `ConnectionError`. Different operations may return different error cases, so it doesn't always make sense to handle the same kinds of errors everywhere. I suggest that we split the `ConnectionError` into smaller error types. For example:

- `ConnectionError` would be returned by `Session::connect`,
- `ExecuteError` would be returned by `Session::query`, `Session::execute`, `Session::batch`, and probably the `QueryIterator`,
- `PrepareError` - by `Session::prepare`,
- etc.

This particular split-up is a suggestion of course, so if you have a better idea for how those should be split or know better names - go for it.

Next, the `ConnectionError` has _lots_ of variants, and I think we could aggregate some of them. Similar kinds of errors could be put into a separate error type, and we would have only one variant in our user-facing error types which encapsulate it.

The rule of thumb for deciding when the errors should be separate or not is: does user want to react to those errors in a different way? For example, for the `ExecuteError`, we could have the following variants:

- Internal driver errors - unmet invariants in the drivers, protocol errors, etc. This would include `UnexpectedFrame`, `UnexpectedResponse`, `RefreshingFailedOnEveryConnections`, `LocalTokensEmptyOnNodes`, `PoolLockPoisoned`, `UpdaterError`, `FatalConnectionError`, `ParseError`, `TopologyError`, `FrameError`. We could move those cases to a new type: `InternalDriverError`.
- I/O errors - those probably should be separate from internal driver errors, because sometimes we would like to retry an operation if an I/O error occurs.
- Errors returned from the server (they are documented [here](https://github.com/apache/cassandra/blob/trunk/doc/native_protocol_v4.spec))

So, for example, the `ExecuteError` should look more like this:

```rust
pub enum ExecuteError {
    InternalDriverError(#[from] InternalDriverError),
    IOError(#[from] std::io::Error),
    ErrorResponse(#[from] ErrorResponse),
}

pub enum ErrorResponse {
    // All of those are errors from the protocol spec
    ServerError(...),
    ProtocolError(...),
    AuthenticationError(...),
    UnavailableException(...),
    Overloaded(...),
    // ... and so on
}
```

----

... I hope that what I described here is not too much work for this PR. If it is, then we can probably do this re-arranging in followup PRs.

@psarna WDYT?

**Comment by psarna** [comment]
Let's rearrange now - it's better to get the design right from the start, so people don't have to rebase later. I agree that the most important thing is to separate errors which can be treated differently by the client - e.g. some of them can be retried (failing to read data from a node) or even ignored, while others are unrecoverable (e.g. internal driver errors).

**Comment by psarna** [comment]
A general remark about the workflow: please avoid merging from scylladb/main to the branch will pull request - it generates weird history and makes it much harder to squash/rebase/cherry-pick the change later. Instead, in order to be up to date, rebase your branch on top of scylladb/main and then push:
```bash
git pull --rebase origin/main
```

File: scylla/src/transport/transport_errors.rs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    #[error("system.local `tokens` query returned no columns")]
+    LocalTokensNoColumnsReturned,
+}
+
+#[derive(Error, Debug)]
+pub enum InternalDriverError {
+    #[error("Unexpected response!")]
+    UnexpectedResponse,
+    #[error("refreshing failed on every connection, {0}")]
+    RefreshingFailedOnEveryConnections(String),
```
Comment (psarna) [inline scylla/src/transport/transport_errors.rs:40]: typo: RefreshingFailedOnEveryConnections -> RefreshingFailedOnEveryConnection

File: scylla/src/transport/transport_errors.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    #[error("Pool Lock Poisoned: {0}")]
+    PoolLockPoisoned(String),
+    #[error("Upadter Crashed: {0}")]
+    UpdaterError(String),
+    #[error("fatal error, broken invariant: no connections available")]
+    FatalConnectionError,
+    #[error("Length of provided values ({0}) must be equal to number of batch statements ({1})")]
+    ValueLenMismatch(usize, usize),
+    #[error("Response is an error message: code: {0} message: {1}")]
+    ErrorMsg(i32, String),
```
[the remark this answers is not in the document]
Comment (psarna) [reply scylla/src/transport/transport_errors.rs:70]: I agree, this comment still stands.

**Pull request merged**